### PR TITLE
LPD-30790 Let AuthVerifiers be prioritized

### PR DESCRIPTION
**Pull request by marianoalvarosaiz**

### CODE REVIEW

**Comment by marianoalvarosaiz** [comment]
Only the former AuthVerifier have a service.ranking to ensure same order as portal.property.

Rest of AuthVerifier will have the default service.ranking, so they will be executed at the end, which doesn't affect the functionality because the order is mainly relevant for BasicAuth, others will keep trying to auth through the pipeline until a suitable one is found.

**Comment by alvarosaugarlr** [comment]
ci:test:stable

**Comment by alvarosaugarlr** [comment]
ci:tst:security

**Comment by alvarosaugarlr** [comment]
ci:test:relevant

**Comment by alvarosaugarlr** [comment]
ci:test:relevant

**Comment by alvarosaugarlr** [comment]
ci:test:security

File: portal-impl/src/com/liferay/portal/security/auth/AuthVerifierPipeline.java [hunk 1]
```diff
@@ -212,6 +222,32 @@ private synchronized void _removeAuthVerifierConfiguration(
 	private volatile URLPatternMapper<List<AuthVerifierConfiguration>>
 		_includeURLPatternMapper;
 
+	private static class AuthVerifierConfigurationCompartor
+		implements Comparator<AuthVerifierConfiguration> {
```
Comment (stian-sigvartsen) [inline portal-impl/src/com/liferay/portal/security/auth/AuthVerifierPipeline.java:226]: Type in `Compartor`. Should be `Comparator`

**Comment by stian-sigvartsen** [comment]
In offline conversation with @marianoalvarosaiz this change is consistent with the docs that when BASIC or Digest auth is configured to issue challenges, then it also means that BASIC / Digest auth is required. So other AuthVerifiers configured for the path become ignored.

@alvarosaugarlr with this in mind all that remains now is to review the implementation of this agreed behaviour.

File: portal-impl/src/com/liferay/portal/security/auth/AuthVerifierPipeline.java (hunk 1 above)
Comment (marianoalvarosaiz) [reply portal-impl/src/com/liferay/portal/security/auth/AuthVerifierPipeline.java:226]: Good catch!

**Comment by marianoalvarosaiz** [comment]
I've force pushed the branch fixing (only) the typo Stian detected.

**Comment by alvarosaugarlr** [comment]
Taking in mind the conversation between @marianoalvarosaiz and @stian-sigvartsen I have tested the case included in the description and more cases to be sure that the order to register the services is what is defined in the service.ranking. For example I have debugged to check the order.

Unit test to cover these changes:
* [first](https://github.com/liferay-appsec/liferay-portal/pull/1840/commits/cc906897a01024ef54aa8ff2d13199e9465a17ba) to prepare tests to be able to work with several AuthVerifiers to check which one is executed first
* testVerifyRequestConsidersAuthVerifiersOrder - Check that order (defined by service.ranking) is taken into account in the pipeline.
Pass in local and the ci

**Comment by rafaprax** [comment]
resent here https://github.com/liferay-appsec/liferay-portal/pull/1909

**Comment by rafaprax** [comment]
ci:close

**Pull request closed without merging**